File: App/Sources/Features/Members/MembersJoinedText.swift
import Foundation
import TeamTasksCore

/// « Membre depuis le 14 septembre » under a member's name: the day only (the time of joining does not matter), with
/// its article, short enough for one caption line.
enum MembersJoinedText {
    /// « Membre depuis aujourd’hui », « … hier », « … le 14 septembre », « … le 1er décembre 2025 » (the year when it
    /// is not the current one).
    static func sentence(joinedAt: Date, now: Date, calendar: Calendar) -> String {
        "Membre depuis \(day(joinedAt, now: now, calendar: calendar))"
    }

    private static func day(_ date: Date, now: Date, calendar: Calendar) -> String {
        let formatter = FrenchDateFormatter(timeZone: calendar.timeZone)
        switch formatter.dayOffset(of: date, from: now) {
        case 0...:
            // Today, or « tomorrow » from a clock set late: never « depuis demain ».
            return "aujourd’hui"
        case -1:
            return "hier"
        default:
            let components = formatter.calendar.dateComponents([.day, .month, .year], from: date)
            let dayNumber = components.day ?? 1
            let month = FrenchDateFormatter.monthNames[((components.month ?? 1) - 1 + 12) % 12]
            var text = "le \(dayNumber == 1 ? "1er" : "\(dayNumber)") \(month)"
            if let year = components.year, year != formatter.calendar.component(.year, from: now) {
                text += " \(year)"
            }
            return text
        }
    }
}

File: android/app/src/main/kotlin/io/github/notkanaa/equipe/ui/members/MembersJoinedText.kt
package io.github.notkanaa.equipe.ui.members

import io.github.notkanaa.equipe.core.AppCalendar
import io.github.notkanaa.equipe.core.logic.FrenchDateFormatter
import java.time.Instant

// Port of App/Sources/Features/Members/MembersJoinedText.swift.

/**
 * « Membre depuis le 14 septembre » under a member's name: the day only (the time of joining does not matter), with
 * its article, short enough for one caption line.
 */
internal object MembersJoinedText {
    /**
     * « Membre depuis aujourd’hui », « … hier », « … le 14 septembre », « … le 1er décembre 2025 » (the year when it
     * is not the current one).
     */
    fun sentence(joinedAt: Instant, now: Instant, calendar: AppCalendar): String =
        "Membre depuis ${day(joinedAt, now, calendar)}"

    private fun day(date: Instant, now: Instant, calendar: AppCalendar): String {
        val formatter = FrenchDateFormatter.of(calendar)
        val offset = formatter.dayOffset(date, now)
        return when {
            // Today, or « tomorrow » from a clock set late: never « depuis demain ».
            offset >= 0 -> "aujourd’hui"
            offset == -1 -> "hier"
            else -> {
                val day = calendar.localDate(date)
                val dayNumber = if (day.dayOfMonth == 1) "1er" else day.dayOfMonth.toString()
                val month = FrenchDateFormatter.MONTH_NAMES[day.monthValue - 1]
                val text = "le $dayNumber $month"
                if (day.year != calendar.localDate(now).year) "$text ${day.year}" else text
            }
        }
    }
}
